### DISCUSSION TITLE
Advanced Foundey -Defi - Stable Coin Creation.

### DISCUSSION THREAD

**Discussion by 18Abhinav07** [Cyfrin/foundry-full-course-cu#2790] [Q&A]
In the DSCEngine.sol :

    function _redeemCollateral(
        address tokenCollateralAddress,
        uint256 amountCollateral,
        address from,
        address to
    )
        private
    {
        s_collateralDeposited[from][tokenCollateralAddress] -= amountCollateral;
        emit CollateralRedeemed(from, to, tokenCollateralAddress, amountCollateral);
        bool success = IERC20(tokenCollateralAddress).transfer(to, amountCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
    
    TEST FUNCTION:

  function testRedeemCollateralWorks() public depositedCollateralAndMintedDsc {
        uint256 userBalance = dsc.balanceOf(USER);
        assertEq(userBalance, amountToMint);

        vm.startPrank(USER);
        dscEngine.redeemCollateral(weth, amountToMint);
        vm.stopPrank();

        uint256 collateralBalance = dscEngine.getCollateralBalanceOfUser(USER, weth);
        assertEq(collateralBalance, amountCollateral - dscEngine.getTokenAmountFromUsd(weth, amountToMint));
    }

the amountToMint is 100 ether as in Patrick's code but when we deposit collateral we deposit 10 ether worth and now in the redeem function there is a direct subtraction without any conversion and check. I have some code that I improvised and it works fine. 

MY CODE:

function _redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral, address from, address to)
        private
    {
        //-------------------------------------------------------
        //    s_collateralDeposited[from][tokenCollateralAddress] -= amountCollateral;
        //     emit CollateralRedeemed(from, to, tokenCollateralAddress, amountCollateral);
        //     bool success = IERC20(tokenCollateralAddress).transfer(to, amountCollateral);
        //     if (!success) {
        //         revert DSCEngine__TransferFailed();
        //     }
        //-------------------------------------------------------

        // This wont work beacuse amountCollateral is sent in wei here 100 ether is sent as amount to mint then to redeem but in our test we deposit 10 ether as collateral now we are subtracting 100 ether from 10 ether which is not possible and gives an underflow error.

        // Also sending that collateral amount to the user is not possible as we dont have that much collateral deposited in the contract.

        // @ dev This is due to the direct use of ether for input, we should make something to adapt this conversion by effectively using dollars as input and eth only as collateral value.

        // we need to get the eth value of that redeem amount as that effectively signifies the amount in USD

        // @dev this system is confusing and needs to be revamped.

        uint256 tokenAmountOfCollateral = getTokenAmountFromUsd(tokenCollateralAddress, amountCollateral);
        if (s_collateralDeposited[from][tokenCollateralAddress] < tokenAmountOfCollateral) {
            revert DSCEngine__NotEnoughCollateral();
        }

        s_collateralDeposited[from][tokenCollateralAddress] -= tokenAmountOfCollateral;
        console.log("Left Collateral", s_collateralDeposited[from][tokenCollateralAddress]);
        emit CollateralRedeemed(from, to, tokenCollateralAddress, tokenAmountOfCollateral);
        bool success = IERC20(tokenCollateralAddress).transfer(to, tokenAmountOfCollateral);
        if (!success) {
            revert DSCEngine__TransferFailed();
        }
    }
    
    I want to know if this really is a bug or am I missing something here.
    Thanks

**Comment by EngrPips** [comment]
Hello @18Abhinav07, I am trying to comprehend this.

**Reply by EngrPips** [reply]
Hello @18Abhinav07, But when we are depositing to collateral, we didn't convert the collateral to USD, so why do you expect that when we want to redeem it, we are supposed to convert it to USD. Below is my `depositCollateral` function from the course.


```js
function depositCollateral(address _address_of_collateral_token, uint256 _amount_to_deposit)
        public
        more_than_zero(_amount_to_deposit)
        token_need_to_be_allowed(_address_of_collateral_token)
        nonReentrant
    {
        s_user_to_amount_deposited_by_user[msg.sender][_address_of_collateral_token] += _amount_to_deposit;
        emit collateral_deposited(msg.sender, _address_of_collateral_token, _amount_to_deposit);

        bool success = IERC20(_address_of_collateral_token).transferFrom(msg.sender, address(this), _amount_to_deposit);
        if (!success) revert DSCEngine__deposit_fail_due_transfer_returning_false();
    }
```
